vms/avm/tx_builders.go

```diff
@@ -0,0 +1,229 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package avm
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/vms/avm/txs"
+	"github.com/ava-labs/avalanchego/vms/components/avax"
+	"github.com/ava-labs/avalanchego/vms/components/verify"
+	"github.com/ava-labs/avalanchego/vms/secp256k1fx"
+	"github.com/ava-labs/avalanchego/wallet/chain/x/signer"
+	"github.com/ava-labs/avalanchego/wallet/subnet/primary/common"
+
+	walletbuilder "github.com/ava-labs/avalanchego/wallet/chain/x/builder"
+)
+
+type txBuilderBackend interface {
+	walletbuilder.Backend
+	signer.Backend
+
+	Context() *walletbuilder.Context
+	ResetAddresses(addrs set.Set[ids.ShortID])
+}
+
+func buildCreateAssetTx(
+	backend txBuilderBackend,
+	name, symbol string,
+	denomination byte,
+	initialStates map[uint32][]verify.State,
+	kc *secp256k1fx.Keychain,
+	changeAddr ids.ShortID,
+) (*txs.Tx, ids.ShortID, error) {
+	pBuilder, pSigner := builders(backend, kc)
+
+	utx, err := pBuilder.NewCreateAssetTx(
+		name,
+		symbol,
+		denomination,
+		initialStates,
+		options(changeAddr, nil /*memo*/)...,
+	)
+	if err != nil {
+		return nil, ids.ShortEmpty, fmt.Errorf("failed building base tx: %w", err)
+	}
+
+	tx, err := signer.SignUnsigned(context.Background(), pSigner, utx)
+	if err != nil {
+		return nil, ids.ShortEmpty, err
+	}
+
+	return tx, changeAddr, nil
+}
+
+func buildBaseTx(
+	backend txBuilderBackend,
+	outs []*avax.TransferableOutput,
+	memo []byte,
+	kc *secp256k1fx.Keychain,
+	changeAddr ids.ShortID,
+) (*txs.Tx, ids.ShortID, error) {
+	pBuilder, pSigner := builders(backend, kc)
+
+	utx, err := pBuilder.NewBaseTx(
+		outs,
+		options(changeAddr, memo)...,
+	)
+	if err != nil {
+		return nil, ids.ShortEmpty, fmt.Errorf("failed building base tx: %w", err)
+	}
+
+	tx, err := signer.SignUnsigned(context.Background(), pSigner, utx)
+	if err != nil {
+		return nil, ids.ShortEmpty, err
+	}
+
+	return tx, changeAddr, nil
+}
+
+func mintNFT(
+	backend txBuilderBackend,
+	assetID ids.ID,
+	payload []byte,
+	owners []*secp256k1fx.OutputOwners,
+	kc *secp256k1fx.Keychain,
+	changeAddr ids.ShortID,
+) (*txs.Tx, error) {
+	pBuilder, pSigner := builders(backend, kc)
+
+	utx, err := pBuilder.NewOperationTxMintNFT(
+		assetID,
+		payload,
+		owners,
+		options(changeAddr, nil /*memo*/)...,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed minting NFTs: %w", err)
+	}
+
+	return signer.SignUnsigned(context.Background(), pSigner, utx)
+}
+
+func mintFTs(
+	backend txBuilderBackend,
+	outputs map[ids.ID]*secp256k1fx.TransferOutput,
+	kc *secp256k1fx.Keychain,
+	changeAddr ids.ShortID,
+) (*txs.Tx, error) {
+	pBuilder, pSigner := builders(backend, kc)
```

Comment:
Variable naming seems to imply this is somehow P-Chain related so I suspect this a copy-paste error